File: client/python/modcdp/NativeMessagingUpstreamTransport.py
from __future__ import annotations

import json
import socket
import struct
import sys
import threading
from pathlib import Path
from collections.abc import Mapping
from typing import Any, cast

from .ExtensionInjector import DEFAULT_MODCDP_EXTENSION_ID
from .UpstreamTransport import UpstreamTransport


DEFAULT_NATIVE_MESSAGING_HOST_NAME = "com.modcdp.bridge"
DEFAULT_NATIVE_MESSAGING_WAIT_TIMEOUT_MS = 10_000


class NativeMessagingUpstreamTransport(UpstreamTransport):
    mode = "nativemessaging"
    endpoint_kind = "modcdp_server"

    def __init__(
        self,
        options: Mapping[str, Any] | None = None,
    ) -> None:
        super().__init__()
        normalized_options = dict(options or {})
        self.manifest_path = cast(str | None, normalized_options.get("manifest_path"))
        self.manifest_paths = list(cast(list[str], normalized_options.get("manifest_paths") or []))
        self.include_default_manifest_paths = self.manifest_path is None and not self.manifest_paths
        self.host_name = str(normalized_options.get("host_name") or DEFAULT_NATIVE_MESSAGING_HOST_NAME)
        self.extension_id = str(normalized_options.get("extension_id") or DEFAULT_MODCDP_EXTENSION_ID)
        self.wait_timeout_ms = int(normalized_options.get("wait_timeout_ms") or DEFAULT_NATIVE_MESSAGING_WAIT_TIMEOUT_MS)
        self.socket: socket.socket | None = None
        self.server: socket.socket | None = None
        self.peer_seen = threading.Event()
        self.bound_port: int | None = None
        self.cdp_url: str | None = None
        self.url = ""

    def update(self, config: dict[str, Any] | None = None) -> "NativeMessagingUpstreamTransport":
        config = config or {}
        if "manifest_path" in config:
            self.manifest_path = config.get("manifest_path")
            self.include_default_manifest_paths = self.manifest_path is None
        if "manifest_paths" in config:
            self.manifest_paths = list(config.get("manifest_paths") or [])
            self.include_default_manifest_paths = len(self.manifest_paths) == 0
        self.extension_id = str(config.get("extension_id") or self.extension_id)
        user_data_dir = config.get("user_data_dir")
        if isinstance(user_data_dir, str) and user_data_dir:
            self._set_profile_manifest_paths(user_data_dir)
            if self.bound_port is not None:
                self._install_native_host(self.bound_port)
        cdp_url = config.get("ws_url") or config.get("cdp_url")
        if isinstance(cdp_url, str) and cdp_url:
            self.cdp_url = cdp_url
        return self

    def getServerConfig(self) -> dict[str, Any]:
        return {"loopback_cdp_url": self.cdp_url} if self.cdp_url else {}

    def getInjectorConfig(self) -> dict[str, Any]:
        return {"native_host_name": self.host_name}

    def connect(self) -> None:
        server = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        server.bind(("127.0.0.1", 0))
        server.listen(1)
        self.server = server
        self.bound_port = int(server.getsockname()[1])
        self.url = f"native://{self.host_name}@127.0.0.1:{self.bound_port}"
        self._install_native_host(self.bound_port)
        threading.Thread(target=self._accept_loop, daemon=True).start()

    def send(self, message: dict[str, Any]) -> None:
        if self.socket is None:
            raise RuntimeError(f"No native messaging peer is connected for {self.host_name}.")
        _write_length_prefixed_json(self.socket, message)

    def waitForPeer(self) -> None:
        if not self.peer_seen.wait(self.wait_timeout_ms / 1000):
            raise RuntimeError(f"Timed out waiting {self.wait_timeout_ms}ms for native messaging host {self.host_name}.")

    def close(self) -> None:
        for sock in (self.socket, self.server):
            try:
                if sock is not None:
                    sock.close()
            except Exception:
                pass
        self.socket = None
        self.server = None

    def _accept_loop(self) -> None:
        server = self.server
        if server is None:
            return
        try:
            conn, _ = server.accept()
        except Exception as error:
            self._emit_close(error if isinstance(error, Exception) else Exception(str(error)))
            return
        if self.socket is not None:
            try:
                self.socket.close()
            except Exception:
                pass
        self.socket = conn
        self.peer_seen.set()
        threading.Thread(target=self._read_loop, args=(conn,), daemon=True).start()

    def _read_loop(self, conn: socket.socket) -> None:
        try:
            for message in _read_length_prefixed_json_messages(conn):
                if message.get("type") == "modcdp.native.hello":
                    continue
                self._emit_recv(message)
        except Exception as error:
            if self.socket is conn:
                self.socket = None
                self._emit_close(error if isinstance(error, Exception) else Exception(str(error)))

    def _install_native_host(self, port: int) -> None:
        host_dir = Path.home() / ".modcdp" / "native-messaging"
        host_dir.mkdir(parents=True, exist_ok=True)
        config_path = host_dir / f"{self.host_name}.config.json"
        host_script_path = host_dir / f"{self.host_name}.py"
        host_executable_path = host_dir / f"{self.host_name}.sh"
        config_path.write_text(json.dumps({"host": "127.0.0.1", "port": port}, indent=2) + "\n")
        host_script_path.write_text(_native_host_script(str(config_path)))
        host_executable_path.write_text(f"#!/bin/sh\nexec {json.dumps(sys.executable)} {json.dumps(str(host_script_path))}\n")
        host_executable_path.chmod(0o755)

        manifest_paths = []
        if self.manifest_path:
            manifest_paths.append(self.manifest_path)
        manifest_paths.extend(self.manifest_paths)
        if self.include_default_manifest_paths:
            manifest_paths.extend(_default_native_messaging_manifest_paths(self.host_name))
        manifest = {
            "name": self.host_name,
            "description": "ModCDP Native Messaging bridge",
            "path": str(host_executable_path),
            "type": "stdio",
            "allowed_origins": [f"chrome-extension://{self.extension_id}/"],
        }
        manifest_text = json.dumps(manifest, indent=2) + "\n"
        for manifest_path in manifest_paths:
            path = Path(manifest_path)
            path.parent.mkdir(parents=True, exist_ok=True)
            path.write_text(manifest_text)

    def _set_profile_manifest_paths(self, user_data_dir: str) -> None:
        self.manifest_paths = [
            str(Path(user_data_dir) / "NativeMessagingHosts" / f"{self.host_name}.json"),
            str(Path(user_data_dir) / "Default" / "NativeMessagingHosts" / f"{self.host_name}.json"),
            *self.manifest_paths,
        ]


def _default_native_messaging_manifest_paths(host_name: str) -> list[str]:
    home = str(Path.home())
    if sys.platform == "darwin":
        return [
            f"{home}/Library/Application Support/Google/Chrome/NativeMessagingHosts/{host_name}.json",
            f"{home}/Library/Application Support/Google/Chrome Canary/NativeMessagingHosts/{host_name}.json",
            f"{home}/Library/Application Support/Google/ChromeForTesting/NativeMessagingHosts/{host_name}.json",
            f"{home}/Library/Application Support/Google/Chrome for Testing/NativeMessagingHosts/{host_name}.json",
            f"{home}/Library/Application Support/Google/Chrome SxS/NativeMessagingHosts/{host_name}.json",
            f"{home}/Library/Application Support/Chromium/NativeMessagingHosts/{host_name}.json",
        ]
    if sys.platform.startswith("linux"):
        return [
            f"{home}/.config/google-chrome/NativeMessagingHosts/{host_name}.json",
            f"{home}/.config/google-chrome-for-testing/NativeMessagingHosts/{host_name}.json",
            f"{home}/.config/chromium/NativeMessagingHosts/{host_name}.json",
            f"{home}/.config/chromium-browser/NativeMessagingHosts/{host_name}.json",
        ]
    raise RuntimeError("upstream nativemessaging manifest_path is required on this platform.")


def _write_length_prefixed_json(sock: socket.socket, message: dict[str, Any]) -> None:
    body = json.dumps(message).encode()
    sock.sendall(struct.pack("<I", len(body)) + body)


def _read_exact(sock: socket.socket, length: int) -> bytes:
    chunks: list[bytes] = []
    remaining = length
    while remaining > 0:
        chunk = sock.recv(remaining)
        if not chunk:
            raise RuntimeError("native messaging socket closed")
        chunks.append(chunk)
        remaining -= len(chunk)
    return b"".join(chunks)


def _read_length_prefixed_json_messages(sock: socket.socket):
    while True:
        header = _read_exact(sock, 4)
        length = struct.unpack("<I", header)[0]
        body = _read_exact(sock, length)
        message = json.loads(body.decode())
        if isinstance(message, dict):
            yield message


def _native_host_script(config_path: str) -> str:
    return f"""
import json
import socket
import struct
import sys
import threading

config = json.loads(open({config_path!r}, "r", encoding="utf-8").read())
sock = socket.create_connection((config["host"], config["port"]))

def write_native(message):
    body = json.dumps(message).encode()
    sys.stdout.buffer.write(struct.pack("<I", len(body)) + body)
    sys.stdout.buffer.flush()

def write_tcp(message):
    body = json.dumps(message).encode()
    sock.sendall(struct.pack("<I", len(body)) + body)

def read_exact(source, length):
    chunks = []
    remaining = length
    while remaining > 0:
        chunk = source.read(remaining) if hasattr(source, "read") else source.recv(remaining)
        if not chunk:
            raise SystemExit(0)
        chunks.append(chunk)
        remaining -= len(chunk)
    return b"".join(chunks)

def read_messages(source, on_message):
    while True:
        header = read_exact(source, 4)
        length = struct.unpack("<I", header)[0]
        body = read_exact(source, length)
        on_message(json.loads(body.decode()))

write_tcp({{"type": "modcdp.native.hello", "role": "native-host", "version": 1}})
threading.Thread(target=read_messages, args=(sys.stdin.buffer, write_tcp), daemon=True).start()
read_messages(sock, write_native)
"""

File: bridge/NativeMessagingUpstreamTransport.ts
import { mkdtempSync } from "node:fs";
import { tmpdir } from "node:os";
import path from "node:path";
import type { CdpCommandMessage } from "../types/modcdp.js";
import type { BrowserLaunchOptions } from "./BrowserLauncher.js";
import { UpstreamTransport, type UpstreamTransportConfig } from "./UpstreamTransport.js";

export const DEFAULT_NATIVE_MESSAGING_HOST_NAME = "com.modcdp.bridge";
export const DEFAULT_NATIVE_MESSAGING_WAIT_TIMEOUT_MS = 10_000;
export const DEFAULT_MODCDP_EXTENSION_ID = "mdedooklbnfejodmnhmkdpkaedafkehf";

type NativeMessagingOptions = {
  manifest_path?: string | null;
  manifest_paths?: string[] | null;
  host_name?: string | null;
  extension_id?: string | null;
  wait_timeout_ms?: number;
};

export class NativeMessagingUpstreamTransport extends UpstreamTransport {
  readonly mode = "nativemessaging" as const;
  readonly endpoint_kind = "modcdp_server" as const;
  url = "";
  private server: any = null;
  private socket: any = null;
  private peer_waiters = new Set<() => void>();
  private wait_timeout_ms: number;
  private manifest_path: string | null;
  private manifest_paths: string[];
  private host_name: string;
  private extension_id: string;
  private user_data_dir: string | null = null;
  private cdp_url: string | null = null;

  constructor({
    manifest_path = null,
    manifest_paths = null,
    host_name = DEFAULT_NATIVE_MESSAGING_HOST_NAME,
    extension_id = DEFAULT_MODCDP_EXTENSION_ID,
    wait_timeout_ms = DEFAULT_NATIVE_MESSAGING_WAIT_TIMEOUT_MS,
  }: NativeMessagingOptions = {}) {
    super();
    this.manifest_path = manifest_path;
    this.manifest_paths = manifest_paths ?? [];
    this.host_name = host_name || DEFAULT_NATIVE_MESSAGING_HOST_NAME;
    this.extension_id = extension_id || DEFAULT_MODCDP_EXTENSION_ID;
    this.wait_timeout_ms = wait_timeout_ms;
  }

  update(config: UpstreamTransportConfig = {}) {
    this.manifest_path = config.manifest_path ?? this.manifest_path;
    this.manifest_paths = config.manifest_paths ?? this.manifest_paths;
    this.user_data_dir = config.user_data_dir ?? this.user_data_dir;
    this.cdp_url = config.ws_url ?? config.cdp_url ?? this.cdp_url;
    if (!this.manifest_path && this.user_data_dir) this.setProfileManifestPaths(this.user_data_dir);
    return this;
  }

  getLauncherConfig(): BrowserLaunchOptions {
    if (!this.manifest_path && !this.user_data_dir) {
      this.user_data_dir = mkdtempSync(path.join(tmpdir(), "modcdp-native-profile-"));
      this.setProfileManifestPaths(this.user_data_dir);
      return { user_data_dir: this.user_data_dir, cleanup_user_data_dir: true };
    }
    return this.user_data_dir ? { user_data_dir: this.user_data_dir } : {};
  }

  getServerConfig() {
    return this.cdp_url ? { loopback_cdp_url: this.cdp_url } : {};
  }

  async connect() {
    if (typeof process !== "object" || !process?.versions?.node) {
      throw new Error("upstream.mode=nativemessaging requires Node.");
    }
    const net = await import("node:net");
    const server = net.createServer((socket) => this.accept(socket));
    this.server = server;
    await new Promise<void>((resolve, reject) => {
      server.once("error", reject);
      server.listen(0, "127.0.0.1", () => resolve());
    });
    const address = server.address();
    if (!address || typeof address === "string") throw new Error("Native messaging bridge did not bind a TCP port.");
    this.url = `native://${this.host_name}@127.0.0.1:${address.port}`;
    await this.installNativeHost(address.port);
  }

  send(message: CdpCommandMessage) {
    if (!this.socket || this.socket.destroyed)
      throw new Error(`No native messaging peer is connected for ${this.host_name}.`);
    writeLengthPrefixedJSON(this.socket, message);
  }

  async waitForPeer() {
    if (this.socket && !this.socket.destroyed) return;
    await new Promise<void>((resolve, reject) => {
      const timeout = setTimeout(
        () =>
          reject(new Error(`Timed out waiting ${this.wait_timeout_ms}ms for native messaging host ${this.host_name}.`)),
        this.wait_timeout_ms,
      );
      this.peer_waiters.add(() => {
        clearTimeout(timeout);
        resolve();
      });
    });
  }

  async close() {
    try {
      this.socket?.destroy();
    } catch {}
    this.socket = null;
    if (this.server) await new Promise<void>((resolve) => this.server.close(() => resolve()));
    this.server = null;
  }

  private accept(socket: any) {
    if (this.socket && this.socket !== socket) {
      try {
        this.socket.destroy();
      } catch {}
    }
    this.socket = socket;
    const readMessage = createLengthPrefixedJSONReader((message) => {
      if (message?.type === "modcdp.native.hello") return;
      this.parseAndEmitRecv(JSON.stringify(message));
    });
    socket.on("data", readMessage);
    socket.on("close", () => {
      if (this.socket !== socket) return;
      this.socket = null;
      this.emitClose(new Error("Native messaging host disconnected"));
    });
    socket.on("error", () => {
      if (this.socket !== socket) return;
      this.socket = null;
      this.emitClose(new Error("Native messaging host error"));
    });
    for (const waiter of this.peer_waiters) waiter();
    this.peer_waiters.clear();
  }

  private async installNativeHost(port: number) {
    const fs = await import("node:fs");
    const os = await import("node:os");
    const path = await import("node:path");
    const hostDir = path.join(os.homedir(), ".modcdp", "native-messaging");
    fs.mkdirSync(hostDir, { recursive: true });

    const configPath = path.join(hostDir, `${this.host_name}.config.json`);
    const hostScriptPath = path.join(hostDir, `${this.host_name}.mjs`);
    const hostExecutablePath = path.join(hostDir, `${this.host_name}.sh`);
    fs.writeFileSync(configPath, JSON.stringify({ host: "127.0.0.1", port }, null, 2));
    fs.writeFileSync(hostScriptPath, nativeHostScript(configPath));
    fs.writeFileSync(
      hostExecutablePath,
      `#!/bin/sh\nexec ${JSON.stringify(process.execPath)} ${JSON.stringify(hostScriptPath)}\n`,
    );
    fs.chmodSync(hostExecutablePath, 0o755);

    const manifestPaths =
      this.manifest_path || this.manifest_paths.length > 0
        ? [...(this.manifest_path ? [this.manifest_path] : []), ...this.manifest_paths]
        : defaultNativeMessagingManifestPaths(this.host_name, os.homedir());
    const manifest = JSON.stringify(
      {
        name: this.host_name,
        description: "ModCDP Native Messaging bridge",
        path: hostExecutablePath,
        type: "stdio",
        allowed_origins: [`chrome-extension://${this.extension_id}/`],
      },
      null,
      2,
    );
    for (const manifestPath of manifestPaths) {
      fs.mkdirSync(path.dirname(manifestPath), { recursive: true });
      fs.writeFileSync(manifestPath, manifest);
    }
  }

  private setProfileManifestPaths(user_data_dir: string) {
    this.manifest_path = path.join(user_data_dir, "NativeMessagingHosts", `${this.host_name}.json`);
    this.manifest_paths = [
      path.join(user_data_dir, "Default", "NativeMessagingHosts", `${this.host_name}.json`),
      ...this.manifest_paths,
    ];
  }
}

function defaultNativeMessagingManifestPaths(host_name: string, home: string) {
  if (process.platform === "darwin") {
    return [
      `${home}/Library/Application Support/Google/Chrome/NativeMessagingHosts/${host_name}.json`,
      `${home}/Library/Application Support/Google/Chrome Canary/NativeMessagingHosts/${host_name}.json`,
      `${home}/Library/Application Support/Google/ChromeForTesting/NativeMessagingHosts/${host_name}.json`,
      `${home}/Library/Application Support/Google/Chrome for Testing/NativeMessagingHosts/${host_name}.json`,
      `${home}/Library/Application Support/Google/Chrome SxS/NativeMessagingHosts/${host_name}.json`,
      `${home}/Library/Application Support/Chromium/NativeMessagingHosts/${host_name}.json`,
    ];
  }
  if (process.platform === "linux") {
    return [
      `${home}/.config/google-chrome/NativeMessagingHosts/${host_name}.json`,
      `${home}/.config/google-chrome-for-testing/NativeMessagingHosts/${host_name}.json`,
      `${home}/.config/chromium/NativeMessagingHosts/${host_name}.json`,
      `${home}/.config/chromium-browser/NativeMessagingHosts/${host_name}.json`,
    ];
  }
  throw new Error("upstream-nativemessaging-manifest is required on this platform.");
}

function writeLengthPrefixedJSON(stream: { write: (chunk: Buffer) => void }, message: unknown) {
  const body = Buffer.from(JSON.stringify(message), "utf8");
  const header = Buffer.alloc(4);
  header.writeUInt32LE(body.length, 0);
  stream.write(Buffer.concat([header, body]));
}

function createLengthPrefixedJSONReader(onMessage: (message: any) => void) {
  let buffer = Buffer.alloc(0);
  return (chunk: Buffer) => {
    buffer = Buffer.concat([buffer, chunk]);
    while (buffer.length >= 4) {
      const length = buffer.readUInt32LE(0);
      if (buffer.length < length + 4) return;
      const body = buffer.subarray(4, 4 + length);
      buffer = buffer.subarray(4 + length);
      try {
        onMessage(JSON.parse(body.toString("utf8")));
      } catch {}
    }
  };
}

function nativeHostScript(configPath: string) {
  return `
import fs from "node:fs";
import net from "node:net";

const config = JSON.parse(fs.readFileSync(${JSON.stringify(configPath)}, "utf8"));
const socket = net.createConnection({ host: config.host, port: config.port }, () => {
  writeTCP({ type: "modcdp.native.hello", role: "native-host", version: 1 });
});
let stdinBuffer = Buffer.alloc(0);
let socketBuffer = Buffer.alloc(0);

process.stdin.on("data", (chunk) => {
  stdinBuffer = Buffer.concat([stdinBuffer, chunk]);
  stdinBuffer = readMessages(stdinBuffer, (message) => {
    writeTCP(message);
  });
});
socket.on("data", (chunk) => {
  socketBuffer = Buffer.concat([socketBuffer, chunk]);
  socketBuffer = readMessages(socketBuffer, (message) => {
    writeNative(message);
  });
});
socket.on("close", () => process.exit(0));
socket.on("error", () => process.exit(1));

function readMessages(buffer, onRecv) {
  while (buffer.length >= 4) {
    const length = buffer.readUInt32LE(0);
    if (buffer.length < length + 4) return buffer;
    const body = buffer.subarray(4, 4 + length);
    buffer = buffer.subarray(4 + length);
    try {
      onRecv(JSON.parse(body.toString("utf8")));
    } catch {}
  }
  return buffer;
}

function writeNative(message) {
  const body = Buffer.from(JSON.stringify(message), "utf8");
  const header = Buffer.alloc(4);
  header.writeUInt32LE(body.length, 0);
  process.stdout.write(Buffer.concat([header, body]));
}

function writeTCP(message) {
  const body = Buffer.from(JSON.stringify(message), "utf8");
  const header = Buffer.alloc(4);
  header.writeUInt32LE(body.length, 0);
  socket.write(Buffer.concat([header, body]));
}
`;
}
